modules/navier_stokes/src/actions/INSAction.C

```diff
@@ -0,0 +1,928 @@
+    params.set<bool>("integrate_p_by_parts") = getParam<bool>("integrate_p_by_parts");
+    params.set<bool>("supg") = getParam<bool>("supg");
+
+    for (unsigned int component = 0; component < _dim; ++component)
+    {
+      params.set<NonlinearVariableName>("variable") = momentums[component];
+      params.set<unsigned int>("component") = component;
+      _problem->addKernel(kernel_type, momentums[component] + std::string("_if"), params);
+    }
+  }
+}
+
+void
+INSAction::addINSTemperature()
+{
+  if (_use_ad)
+  {
+    {
+      const std::string kernel_type = "INSADTemperatureAdvection";
+      InputParameters params = _factory.getValidParams(kernel_type);
+      params.set<NonlinearVariableName>("variable") = _temperature_variable_name;
+      params.set<std::vector<VariableName>>("velocity") = {NS::velocity};
+      if (_blocks.size() > 0)
+        params.set<std::vector<SubdomainName>>("block") = _blocks;
+      _problem->addKernel(kernel_type, "ins_temperature_convection", params);
+    }
+    {
+      const std::string kernel_type = "ADHeatConduction";
+      InputParameters params = _factory.getValidParams(kernel_type);
+      params.set<NonlinearVariableName>("variable") = _temperature_variable_name;
+      params.set<MaterialPropertyName>("thermal_conductivity") =
+          getParam<MaterialPropertyName>("thermal_conductivity_name");
+      if (_blocks.size() > 0)
+        params.set<std::vector<SubdomainName>>("block") = _blocks;
+      _problem->addKernel(kernel_type, "ins_temperature_conduction", params);
+    }
+  }
+  else
+  {
+    const std::string kernel_type = "INSTemperature";
+    InputParameters params = _factory.getValidParams(kernel_type);
+    params.set<NonlinearVariableName>("variable") = _temperature_variable_name;
+    params.set<CoupledName>("u") = {NS::velocity_x};
+    if (_dim >= 2)
+      params.set<CoupledName>("v") = {NS::velocity_y};
+    if (_dim >= 3)
+      params.set<CoupledName>("w") = {NS::velocity_z};
+    params.set<MaterialPropertyName>("k_name") =
+        getParam<MaterialPropertyName>("thermal_conductivity_name");
+    params.set<MaterialPropertyName>("rho_name") = getParam<MaterialPropertyName>("density_name");
+    params.set<MaterialPropertyName>("cp_name") =
+        getParam<MaterialPropertyName>("specific_heat_name");
+    if (_blocks.size() > 0)
+      params.set<std::vector<SubdomainName>>("block") = _blocks;
+    _problem->addKernel(kernel_type, "ins_temperature", params);
+  }
+}
+
+void
+INSAction::addINSVelocityBC()
+{
+  const static std::string momentums[3] = {NS::velocity_x, NS::velocity_y, NS::velocity_z};
+  for (unsigned int i = 0; i < _velocity_boundary.size(); ++i)
+  {
+    if (_use_ad)
+    {
+      InputParameters params = _factory.getValidParams("VectorFunctionDirichletBC");
+
+      {
+        const FunctionName funcx = _velocity_function[i * _dim];
+        if (funcx == "NA")
+          mooseError("AD does not support partial velocity BC");
```

Comment:
`ADNodalBC` does now actually support partial specification of vector variable Dirichlet BCs via the `set_x_comp`, `set_y_comp`, and `set_z_comp` parameters. There is an example of using it in INS in `navier_stokes/test/tests/ins/RZ_cone/ad_rz_cone_by_parts.i` (and similar input files)

Reply:
Good to know this!